### PR TITLE
ML-3527: SQL reserved words in query

### PR DESCRIPTION
**Pull request by davesh0812**
Ability to use all sql reserved word as entity or feature when using `sql_driver`

tested only with mysql and sqlite

### CODE REVIEW

File: storey/sql_driver.py [hunk 1]
```diff
@@ -115,15 +115,23 @@ def _get_where_clause(self, key, sql_table):
...
+                where_clause += (
+                    f'{sql_table.name}.[{self._primary_key[i]}]="{key[i]}"'
+                    if "mysql" not in sql_table.dialect_options
+                    else f'{sql_table.name}.{self._primary_key[i]}="{key[i]}"'
+                )
             else:
-                where_clause += f"{self._primary_key[i]}={key[i]}"
+                where_clause += (
+                    f"{sql_table.name}.[{self._primary_key[i]}]={key[i]}"
+                    if "mysql" not in sql_table.dialect_options
```
Comment (tomerm-iguazio) [inline storey/sql_driver.py:126]: [column] will not work in other sql dbs - such as Postgres....

File: storey/sql_driver.py [hunk 2]
```diff
@@ -115,15 +115,23 @@ def _get_where_clause(self, key, sql_table):
...
+                    else f"{sql_table.name}.{self._primary_key[i]}={key[i]}"
+                )
         return where_clause
 
     def _update_by_key(self, key, data, table):
         where_clause = self._get_where_clause(key, table)
-        update_clause = " ,".join([f'{key}="{value}"' for key, value in data.items() if key not in self._primary_key])
-        sql_statement = f"UPDATE {table} SET {update_clause} where {where_clause}"
+        update_clause = " ,".join([f'[{key}]="{value}"' for key, value in data.items() if key not in self._primary_key])
```
Comment (tomerm-iguazio) [inline storey/sql_driver.py]: Again, using [] for column's name is problematic in some sql dbs.

File: integration/test_flow_integration.py [hunk 3]
```diff
@@ -144,10 +144,10 @@ def _get_sql_by_key_all_attrs(
             if i != 0:
                 where_statement += " and "
             if sql_table.columns[key_name[i]].type.python_type == str:
-                where_statement += f'{key_name[i]}="{key[i]}"'
+                where_statement += f'test_table.[{key_name[i]}]="{key[i]}"'
             else:
-                where_statement += f"{key_name[i]}={key[i]}"
-    query = f"SELECT * FROM {sql_table} where ({where_statement})"
+                where_statement += f"test_table.[{key_name[i]}]={key[i]}"
```
Comment (tomerm-iguazio) [inline integration/test_flow_integration.py]: Which sql db are you using?
Not all of them support '[]' to describe column name

File: storey/sql_driver.py [hunk 4]
```diff
@@ -115,15 +115,23 @@ def _get_where_clause(self, key, sql_table):
             if i != 0:
                 where_clause += " and "
             if sql_table.columns[self._primary_key[i]].type.python_type == str:
-                where_clause += f'{self._primary_key[i]}="{key[i]}"'
+                where_clause += (
+                    f'{sql_table.name}.[{self._primary_key[i]}]="{key[i]}"'
```
Comment (tomerm-iguazio) [inline storey/sql_driver.py]: Good, please note that if you use alias - you must use it's new name later on :)

Comment (davesh0812) [reply storey/sql_driver.py]: I am using it's new name no?

File: integration/test_flow_integration.py (hunk 3 above)
Comment (tomerm-iguazio) [reply integration/test_flow_integration.py]: you need to change here to by the relevant db- [] for mysql and "" for non-mysql.

File: storey/sql_driver.py (hunk 4 above)
Comment (tomerm-iguazio) [reply storey/sql_driver.py]: yea all good:)

File: storey/sql_driver.py [hunk 5]
```diff
@@ -115,15 +115,30 @@ def _get_where_clause(self, key, sql_table):
...
+                )
         return where_clause
 
     def _update_by_key(self, key, data, table):
         where_clause = self._get_where_clause(key, table)
-        update_clause = " ,".join([f'{key}="{value}"' for key, value in data.items() if key not in self._primary_key])
-        sql_statement = f"UPDATE {table} SET {update_clause} where {where_clause}"
+        if "mysql" not in table.dialect_options:
+            update_clause = " ,".join(
+                [f"""\"{key}\"=\"{value}\"""" for key, value in data.items() if key not in self._primary_key]
```
Comment (tomerm-iguazio) [inline storey/sql_driver.py]: why are using this \?

I prefer to use raw string:
fr"my_query...."

**Review by tomerm-iguazio — APPROVED** [review]

File: integration/test_flow_integration.py [hunk 6]
```diff
@@ -144,10 +144,10 @@ def _get_sql_by_key_all_attrs(
             if i != 0:
                 where_statement += " and "
             if sql_table.columns[key_name[i]].type.python_type == str:
-                where_statement += f'{key_name[i]}="{key[i]}"'
+                where_statement += rf'test_table."{key_name[i]}"="{key[i]}"'
```
Comment (gtopper) [inline integration/test_flow_integration.py]:
```suggestion
                where_statement += f'test_table."{key_name[i]}"="{key[i]}"'
```

File: integration/test_flow_integration.py [hunk 7]
```diff
@@ -144,10 +144,10 @@ def _get_sql_by_key_all_attrs(
...
             else:
-                where_statement += f"{key_name[i]}={key[i]}"
-    query = f"SELECT * FROM {sql_table} where ({where_statement})"
+                where_statement += rf'test_table."{key_name[i]}"={key[i]}'
```
Comment (gtopper) [inline integration/test_flow_integration.py]:
```suggestion
                where_statement += f'test_table."{key_name[i]}"={key[i]}'
```

File: integration/test_flow_integration.py [hunk 8]
```diff
@@ -144,10 +144,10 @@ def _get_sql_by_key_all_attrs(
...
+    query = rf"SELECT * FROM {sql_table.name} as test_table where {where_statement}"
```
Comment (gtopper) [inline integration/test_flow_integration.py]:
```suggestion
    query = f"SELECT * FROM {sql_table.name} as test_table where {where_statement}"
```

File: storey/sql_driver.py [hunk 9]
```diff
@@ -88,15 +88,15 @@ async def close(self):
 
     async def _get_all_fields(self, key, table):
         where_clause = self._get_where_clause(key, table)
-        query = f"SELECT * FROM {table} where {where_clause}"
+        query = rf"SELECT * FROM {table} where {where_clause}"
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
        query = f"SELECT * FROM {table} where {where_clause}"
```

File: storey/sql_driver.py [hunk 10]
```diff
@@ -88,15 +88,15 @@ async def close(self):
...
         results = pd.read_sql(query, con=self._sql_connection, parse_dates=self._time_fields).to_dict(orient="records")
 
         return results[0]
 
     async def _get_specific_fields(self, key: str, table, attributes: List[str]):
         where_clause = self._get_where_clause(key, table)
         try:
-            query = f"SELECT {','.join(attributes)} FROM {table} where {where_clause}"
+            query = rf"SELECT {','.join(attributes)} FROM {table} as {table.name} where {where_clause}"
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
            query = f"SELECT {','.join(attributes)} FROM {table} as {table.name} where {where_clause}"
```

File: storey/sql_driver.py [hunk 11]
```diff
@@ -115,15 +115,30 @@ def _get_where_clause(self, key, sql_table):
...
-                where_clause += f'{self._primary_key[i]}="{key[i]}"'
+                where_clause += (
+                    rf'{sql_table.name}."{self._primary_key[i]}"="{key[i]}"'
+                    if "mysql" not in sql_table.dialect_options
+                    else rf'{sql_table.name}.{self._primary_key[i]}="{key[i]}"'
+                )
             else:
-                where_clause += f"{self._primary_key[i]}={key[i]}"
+                where_clause += (
+                    rf"""{sql_table.name}."{self._primary_key[i]}"={key[i]}"""
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
                    f'{sql_table.name}."{self._primary_key[i]}"={key[i]}'
```

File: storey/sql_driver.py [hunk 12]
```diff
@@ -115,15 +115,30 @@ def _get_where_clause(self, key, sql_table):
...
+                    if "mysql" not in sql_table.dialect_options
+                    else rf"{sql_table.name}.{self._primary_key[i]}={key[i]}"
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
                    else f"{sql_table.name}.{self._primary_key[i]}={key[i]}"
```

File: storey/sql_driver.py [hunk 13]
```diff
@@ -115,15 +115,30 @@ def _get_where_clause(self, key, sql_table):
...
-        update_clause = " ,".join([f'{key}="{value}"' for key, value in data.items() if key not in self._primary_key])
-        sql_statement = f"UPDATE {table} SET {update_clause} where {where_clause}"
+        if "mysql" not in table.dialect_options:
+            update_clause = " ,".join(
+                [rf'"{key}"="{value}"' for key, value in data.items() if key not in self._primary_key]
+            )
+        else:
+            update_clause = " ,".join(
+                [rf'[{key}]="{value}"' for key, value in data.items() if key not in self._primary_key]
+            )
```
Comment (gtopper) [inline storey/sql_driver.py]: Too much duplication
```suggestion
        quote = "" if "mysql" in table.dialect_options else '"'
        update_clause = " ,".join(
            [f'[{quote}{key}{quote}]="{value}"' for key, value in data.items() if key not in self._primary_key]
        )
```

File: integration/test_flow_integration.py [hunk 14]
```diff
@@ -1255,20 +1255,20 @@ def test_write_multiple_keys_to_v3io(setup_teardown_test):
 
 
 def test_write_none_time(setup_teardown_test):
-    keys = ["first_name"]
```
Comment (gtopper) [inline integration/test_flow_integration.py:1258]: This looks like it is meant to turn the test into a regression test for ML-3527, but the modified test does not fail when run against development.

Comment (davesh0812) [reply integration/test_flow_integration.py:1258]: @gtopper It is fail to me when I run it against development can you try again?

Comment (gtopper) [reply integration/test_flow_integration.py:1258]: Still passes.

File: storey/sql_driver.py [hunk 15]
```diff
@@ -115,15 +115,29 @@ def _get_where_clause(self, key, sql_table):
...
         return where_clause
 
     def _update_by_key(self, key, data, table):
         where_clause = self._get_where_clause(key, table)
-        update_clause = " ,".join([f'{key}="{value}"' for key, value in data.items() if key not in self._primary_key])
-        sql_statement = f"UPDATE {table} SET {update_clause} where {where_clause}"
+        if "mysql" not in table.dialect_options:
+            left_wrap, right_wrap = "[", "]"
+        else:
+            left_wrap, right_wrap = '"', '"'
```
Comment (gtopper) [inline storey/sql_driver.py]: No need for negation
```suggestion
        if "mysql" in table.dialect_options:
            left_wrap, right_wrap = '"', '"'
        else:
            left_wrap, right_wrap = "[", "]"
```

File: storey/sql_driver.py [hunk 16]
```diff
@@ -115,15 +115,29 @@ def _get_where_clause(self, key, sql_table):
             if i != 0:
                 where_clause += " and "
             if sql_table.columns[self._primary_key[i]].type.python_type == str:
-                where_clause += f'{self._primary_key[i]}="{key[i]}"'
+                where_clause += (
+                    rf'{sql_table.name}."{self._primary_key[i]}"="{key[i]}"'
+                    if "mysql" not in sql_table.dialect_options
+                    else rf'{sql_table.name}.{self._primary_key[i]}="{key[i]}"'
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
                    else f'{sql_table.name}.{self._primary_key[i]}="{key[i]}"'
```

File: storey/sql_driver.py [hunk 17]
```diff
@@ -115,15 +115,29 @@ def _get_where_clause(self, key, sql_table):
             if i != 0:
                 where_clause += " and "
             if sql_table.columns[self._primary_key[i]].type.python_type == str:
-                where_clause += f'{self._primary_key[i]}="{key[i]}"'
+                where_clause += (
+                    rf'{sql_table.name}."{self._primary_key[i]}"="{key[i]}"'
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
                    f'{sql_table.name}."{self._primary_key[i]}"="{key[i]}"'
```

File: storey/sql_driver.py [hunk 18]
```diff
@@ -115,15 +115,29 @@ def _get_where_clause(self, key, sql_table):
...
-        update_clause = " ,".join([f'{key}="{value}"' for key, value in data.items() if key not in self._primary_key])
-        sql_statement = f"UPDATE {table} SET {update_clause} where {where_clause}"
+        if "mysql" not in table.dialect_options:
+            left_wrap, right_wrap = "[", "]"
+        else:
+            left_wrap, right_wrap = '"', '"'
+        update_clause = " ,".join(
+            [f'{left_wrap}{key}{right_wrap}="{value}"' for key, value in data.items() if key not in self._primary_key]
+        )
+        sql_statement = rf"UPDATE {table} as {table.name} SET {update_clause} where {where_clause}"
```
Comment (gtopper) [inline storey/sql_driver.py]:
```suggestion
        sql_statement = f"UPDATE {table} as {table.name} SET {update_clause} where {where_clause}"
```

**Comment by gtopper** [comment]
Have you considered sanitizing the inputs?
Her daughter is named Help I'm trapped in a driver's license factory.

File: integration/test_flow_integration.py (hunk 14 above)
Comment (davesh0812) [reply integration/test_flow_integration.py:1258]: For sql driver?

Comment (davesh0812) [reply integration/test_flow_integration.py:1258]: I got this error
`_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x11dd82bd0>
cursor = <sqlite3.Cursor object at 0x11dbe9f80>
statement = 'SELECT * FROM uUuUAgHotP where (index="yosi")', parameters = ()
context = <sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x11de54a50>

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) near "index": syntax error
E       [SQL: SELECT * FROM uUuUAgHotP where (index="yosi")]
E       (Background on this error at: https://sqlalche.me/e/14/e3q8)

venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py:736: OperationalError`

Comment (gtopper) [reply integration/test_flow_integration.py:1258]: Try with python 3.9? What sqlalchemy version? Mine is 1.4.46.

**Review by gtopper — COMMENTED** [review]
We discussed using query parameters instead of python string substitutions to avoid SQL injection (purposeful or accidental). Currently a string value containing quotes would cause us to build an invalid SQL statement and fail.

**Comment by davesh0812** [comment]
#433

**Pull request closed without merging**